### PR TITLE
load providers depending on features

### PR DESCRIPTION
**Pull request by michaelmattig**
- [ ] I added an entry to [`CHANGELOG.md`](CHANGELOG.md) if knowledge of this change could be valuable to users.

---

Here is a brief summary of what I did:

<TEXT>

### CODE REVIEW

**Comment by michaelmattig** [comment]
[DataProvider nur laden, wenn die Features aktiv sind](https://trello.com/c/d6PTaQ63/1105-dataprovider-nur-laden-wenn-die-features-aktiv-sind)

File: services/src/datasets/add_from_directory.rs [hunk 1]
```diff
@@ -36,9 +37,12 @@ pub async fn add_datasets_from_directory<S: MockableSession, D: DatasetDb<S>>(
...
     }
 
-    let dir = fs::read_dir(file_path);
+    let dir = fs::read_dir(&file_path);
     if dir.is_err() {
-        warn!("Skipped adding datasets from directory because it can't be read");
+        warn!(
+            "Skipped adding datasets from directory `{:?}` because it can't be read",
+            file_path
+        );
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/add_from_directory.rs:45]: Isn't reading from a directory, but cannot access it, more… like an `error!`?

File: services/src/datasets/add_from_directory.rs [hunk 2]
```diff
@@ -36,9 +37,12 @@ pub async fn add_datasets_from_directory<S: MockableSession, D: DatasetDb<S>>(
         Ok(())
     }
 
-    let dir = fs::read_dir(file_path);
+    let dir = fs::read_dir(&file_path);
     if dir.is_err() {
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/add_from_directory.rs:41]:
```Rust
let Ok(dir) = else {
    … output to logs …
    return;
};
```

and then…

File: services/src/datasets/add_from_directory.rs [hunk 3]
```diff
@@ -36,9 +37,12 @@ pub async fn add_datasets_from_directory<S: MockableSession, D: DatasetDb<S>>(
...
-        warn!("Skipped adding datasets from directory because it can't be read");
+        warn!(
+            "Skipped adding datasets from directory `{:?}` because it can't be read",
+            file_path
+        );
         return;
     }
     let dir = dir.expect("checked");
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/add_from_directory.rs:48]: you can make this without expecting

File: services/src/datasets/add_from_directory.rs [hunk 4]
```diff
@@ -74,29 +79,48 @@ pub async fn add_providers_from_directory<D: LayerProviderDb>(db: &mut D, file_p
         Ok(())
     }
 
-    let dir = fs::read_dir(file_path);
-    if dir.is_err() {
-        warn!("Skipped adding providers from directory because it can't be read");
-        return;
-    }
-    let dir = dir.expect("checked");
+    async fn add_from_dir<D: LayerProviderDb>(db: &mut D, file_path: &Path) {
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/add_from_directory.rs:82]:
```suggestion
    async fn add_providers_from_directory<D: LayerProviderDb>(db: &mut D, file_path: &Path) {
```

maybe?

The other comments apply here as well about internals

File: services/src/datasets/add_from_directory.rs [hunk 5]
```diff
@@ -74,29 +79,48 @@ pub async fn add_providers_from_directory<D: LayerProviderDb>(db: &mut D, file_p
...
+                            warn!(
+                                "Skipped adding provider from file `{:?}` error: `{}`",
+                                entry.path(),
+                                e.to_string()
+                            );
+                        }
+                    }
+                }
+                Err(e) => {
+                    // TODO: log
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/add_from_directory.rs:108]:
```suggestion
```

well, you did

File: services/src/datasets/add_from_directory.rs [hunk 6]
```diff
@@ -74,29 +79,48 @@ pub async fn add_providers_from_directory<D: LayerProviderDb>(db: &mut D, file_p
...
-        } else {
-            // TODO: log
-            warn!("Skipped adding provider from directory entry: {:?}", entry);
         }
     }
+
+    add_from_dir(db, &file_path).await;
+
+    #[cfg(feature = "pro")]
+    add_from_dir(db, &file_path.join("pro")).await;
```
Comment (ChristianBeilschmidt) [inline services/src/datasets/add_from_directory.rs:119]: We said our goal was to have only one entry-point for pro.
Can we somehow keep it this way for this place?

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: services/src/datasets/add_from_directory.rs (hunk 6 above)
Comment (michaelmattig) [reply services/src/datasets/add_from_directory.rs:119]: we could duplicte the whole method in pro and only add the "pro" subfolder there.

or accept multiple `file_path`s and put only the base folder in, for non-pro, and the pro-folder additionaly in pro

File: services/src/datasets/add_from_directory.rs (hunk 1 above)
Comment (michaelmattig) [reply services/src/datasets/add_from_directory.rs]: 8ccc34bde0a3e09e5cd4346a9044f2bc7fbd9535

File: services/src/datasets/add_from_directory.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/datasets/add_from_directory.rs]: I applied that new pattern 8ccc34bde0a3e09e5cd4346a9044f2bc7fbd9535

File: services/src/datasets/add_from_directory.rs (hunk 3 above)
Comment (michaelmattig) [reply services/src/datasets/add_from_directory.rs]: I applied the new pattern 8ccc34bde0a3e09e5cd4346a9044f2bc7fbd9535

File: services/src/datasets/add_from_directory.rs (hunk 5 above)
Comment (michaelmattig) [reply services/src/datasets/add_from_directory.rs]: I removed it 8ccc34bde0a3e09e5cd4346a9044f2bc7fbd9535

File: services/src/datasets/add_from_directory.rs (hunk 6 above)
Comment (ChristianBeilschmidt) [reply services/src/datasets/add_from_directory.rs:114]: The last thing sounds better than duplicating everything!

Comment (michaelmattig) [reply services/src/datasets/add_from_directory.rs]: b0a9416f44a392c08b060cd1a4fe013523efc9b6

**Review by ChristianBeilschmidt — APPROVED** [review]

**Comment by ChristianBeilschmidt** [comment]
bors merge

**Pull request merged**